api/config/v1/sharing_test.go

```diff
@@ -0,0 +1,197 @@
+/**
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package v1
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestUnmarshalSharingDefaults(t *testing.T) {
+	testCases := []struct {
+		input           string
+		output          Sharing
+		isErrorExpected bool
+	}{
+		{
+			input:           ``,
+			isErrorExpected: true,
+		},
+		{
+			input: `{}`,
+		},
+		{
+			input: `{
+				"mps": {
+					"resources": [
+						{ "name": "nvidia.com/gpu", "replicas": 10 }
+					]
+				}
+			}`,
+			output: Sharing{
+				MPS: &ReplicatedResources{
+					FailRequestsGreaterThanOne: true,
+					Resources: []ReplicatedResource{
+						{
+							Name:     "nvidia.com/gpu",
+							Replicas: 10,
+							Devices: ReplicatedDevices{
+								All: true,
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			input: `{
+				"mps": {
+					"failRequestsGreaterThanOne": true,
+					"resources": [
+						{ "name": "nvidia.com/gpu", "replicas": 10 }
+					]
+				}
+			}`,
+			output: Sharing{
+				MPS: &ReplicatedResources{
+					FailRequestsGreaterThanOne: true,
+					Resources: []ReplicatedResource{
+						{
+							Name:     "nvidia.com/gpu",
+							Replicas: 10,
+							Devices: ReplicatedDevices{
+								All: true,
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			input: `{
+				"mps": {
+					"failRequestsGreaterThanOne": false,
```

Comment:
Question -- what does it effectively mean if `failRequestsGreaterThanOne=false` and in my pod spec I request multiple `nvidia.com/gpu` resources for my container (both for MPS and time-slicing)? For MPS specifically, does requesting multiple `nvidia.com/gpu` resources increase the amount of compute / memory allocated to the container? Or does it just limit the amount of countable resources available for other containers?

Reply:
I don't even think we should have this option for MPS -- or if we do, it should give you multiple slices, instead of 1 (which we can't do currently because we limit the size of each slice at the control daemon level).

Reply:
The issue is that we don't have control over which devices the kubelet is allocating and cannot guarantee that the devices / slices are on the same device. We can relax this restriction IF there is only a single device this is what the GKE plugin does here: https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/cf2cc6e7a7d0ed2c2d2c2373fcc7fad7233b8be6/pkg/gpu/nvidia/gpusharing/gpusharing.go#L33-L50

Note that they do combine multiple MPS slices since they set the MPS envvars for the container based on the number of requested devices on a single device:  https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/cf2cc6e7a7d0ed2c2d2c2373fcc7fad7233b8be6/pkg/gpu/nvidia/manager.go#L312-L325

The intent of this change was only to ensure that this option is set to `true` by default for MPS, but I'm also happy to not allow this to be changed.